### ISSUE TITLE
Rethink the way how page templates work to support multiple templates per post type

### ISSUE THREAD

**Issue by Mahjouba91** [WordPress/gutenberg#19907]
**Is your feature request related to a problem? Please describe.**
I'm frustrated about how templates currently work in Gutenberg.

You can implement a template through PHP like this:
```php
function register_editor_templates() {
	$post_type_object = get_post_type_object( 'post' );
	if ( null !== $post_type_object ) {
		$post_type_object->template = array(
			array(
				'core/paragraph',
				array(
					'placeholder' => __( 'Write an introduction...' ),
				),
			),
			array( 'core/more' ),
			array(
				'core/paragraph',
				array(
					'placeholder' => __( 'Write more details...' ),
				),
			),
		);
	}
}
add_action( 'init', 'register_editor_templates' );
```

The problem is that you can only assign one template to a post type, and that this template will be automatically loaded at every new post creation.

Then we have the JS way by creating a new block. So it's nice because you can select any template you want for you post type but it's still not ideal. When adding a template through a block, it just loads all blocks of this template in the current post, even if you have already like 5 blocks in your post, it will just add it anywhere you want. so if you make a mistake and choose the wrong template for example, it's tricky, you will end up with a lot of blocks mixed in your content.

**Describe the solution you'd like**
I would like a solution similar to what we had previously with the classic editor. That means, to have a select box in order to choose any template we want so it would change all blocks of the current post.

If the current template param from a post type could be an array of templates and not juste an array of one template, it would do the trick for the API side, and tou will have to change the behaviour in frontend I guess.

I have also raised another issue that would help to have more granular templates https://github.com/WordPress/gutenberg/issues/19874

**Describe alternatives you've considered**
Basically, the two official implementations all have drawbacks, so I can not implement the way I need it for the creators.

**Comment by pwkip** [comment]
I would love to (help) implement this. Just wondering what the API should look like.

I guess it could look as simple as this:
```PHP
$template1 = [
	[
		'core/image'
	],
	[
		'core/paragraph',
		[ 'placeholder' => 'Add a root-level paragraph' ]
	],
];

$template2 = [
	[
		'core/image'
	] 
];

$templates = [
	[
		'templateName' => __('My first template'),
		'template' => $template1
	],
	[
		'templateName' => __('My second template'),
		'template' => $template2
	],
];
```
In the sidebar it could look like this:
![Select Block Template](https://bdwm.be/wp-content/uploads/2020/03/select-block-template.png)

Thoughts?

**Comment by Mahjouba91** [comment]
Hey, thanks for your comment.

Yes, that would be the best option I think

**Comment by LukaszJaro** [comment]
Does the new block pattern feature solve this? Create a pattern and add it to any number of pages?

I have yet to try it but it may be a solution to this.

**Comment by skorasaurus** [comment]
> Does the new block pattern feature solve this? Create a pattern and add it to any number of pages?
> 
> I have yet to try it but it may be a solution to this.

You can add the block pattern to as many or as few pages as you would like. The unfortunate part is that, as of now, you cannot add any block patterns (not even one) to PHP templates - this is being tracked in #21830

I think you'll be able to add them in full site editing templates; but I can't confirm that.

**Comment by maniflames** [comment]
I'm currently working on a theme that needs two different templates for a custom post type and I would love it if I could just register two. I've tried to see how the mentioned issues relate to this but they all seem to lean towards inserting patterns instead. Is the support for multiple templates on the roadmap somewhere? I'd love to still do some logic on the back-end depending on the a template a user has selected if that makes sense.

**Comment by youknowriad** [comment]
Theme can support custom block themes right now. I believe addressing what this issue is about. Also patterns can offer a somewhat similar experience. So there's a lot of possibilities today. We also have starter templates. So I'm closing this as resolved.
